pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go

```diff
@@ -109,7 +109,9 @@ func (s *server) Start() error {
 	api.RegisterRegistrationServer(s.grpc, s)
 	go func() {
 		defer s.wg.Done()
-		s.grpc.Serve(ln)
+		if err = s.grpc.Serve(ln); err != nil {
```

Comment:
I need to dig a bit in the history about why the code is laid out like this, but I think that failing to Serve() should bubble up this error. Logging helps but is likely not enough. Needs to be verified though.

Reply:
I agree, this should buble up the error somehow, I will look into it.

Reply:
Do we need to retry after error?

Reply:
> Do we need to retry after error?

I don't think so. Go uses [accept4](https://linux.die.net/man/2/accept4)and the man page reads

```
For reliable operation the application should detect the network errors defined for the protocol after accept() and treat them like EAGAIN by retrying. In the case of TCP/IP, these are ENETDOWN, EPROTO, ENOPROTOOPT, EHOSTDOWN, ENONET, EHOSTUNREACH, EOPNOTSUPP, and ENETUNREACH. 
```

Looking at Go's internal/poll implementation for unix sockets it already handles the cases that should be retried:

https://github.com/golang/go/blob/5bf8c0c/src/internal/poll/fd_unix.go#L611-L632

Looks like the only errors that can spill out are fatal errors, ie errors that should be handled.

Reply:
I understand that. However, if the kubelet process continues to run, functions related to the device plug-in are unavailable, which may affect the running of the Kubernetes cluster.

Reply:
Yes, I agree with that. Here is an example from the grpc repo https://github.com/grpc/grpc-go/blob/master/examples/features/error_handling/server/main.go#L62-L64, Serve errors should be treated as a fatal error. In that case we should bubble up and stop the program.

Reply:
Bubbling up the error is quite complex since Serve runs in a goroutine and can fail at anytime, most likely not on startup. There are 3 or 4 layers of indirection until the code is finally called in kubelet.

Searching for the references for Serve in the codebase show that it is mostly tests, but when it is used we just print the error:

- dynamic-resource-allocator https://github.com/kubernetes/kubernetes/blob/d59676a/staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/nonblockinggrpcserver.go#L98-L103
- pluginmanager/pluginwatcher https://github.com/kubernetes/kubernetes/blob/f405c79/pkg/kubelet/pluginmanager/pluginwatcher/example_plugin.go#L149-L151

I think just logging the error would be ok, we get some more details instead of it failing silently. Perhaps we should rethink how we orchestrate grpc server code.

Thoughts?

Reply:
I'd be OK with a loud log as starting point

Reply:
@mauri870 please, do the same for https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go#L159, thanks

Reply:
/lgtm